### DISCUSSION TITLE
Port keyszer features missing in xremap

### DISCUSSION THREAD

**Discussion by RedBearAK** [xremap/xremap#613] [Ideas]
@k0kubun 

Hi, I just have a quick question about how your Wayland+GNOME extension presents the application class window attribute. 

I'm integrating your Xremap GNOME Shell extension as an external option to allow another keymapper (keyszer, a fork of xkeysnail) to access the window class and title/name in Wayland. Up until now these other keymappers haven't had access to the window attributes outside of X11/Xorg environments. Your project found a solution to that for a few different Wayland environments, so I'm taking advantage of that. Hope you don't mind. 😸 

When I got this working with another GNOME Shell extension called "Window Calls Extended" that kind of does the same thing, I noticed that sometimes I would see a different window class than expected from working with keyszer/xkeysnail. This turned out to be because keyszer/xkeysnail would pull out the second element from the WM_CLASS pairing, while the Xremap extension and this other GNOME extension seem to be presenting the first element of the pair: 

```sh
WM_CLASS(STRING) = "gnome-terminal-server", "Gnome-terminal"
```

I'm very curious about whether there is some official documented recommendation somewhere that led to these GNOME extensions only showing the first element of this set of two strings that WM_CLASS returns when you use a tool like `xprop` or the Python Xlib library. 

Was this an internal decision when you made the Xremap GNOME extension, or does the GNOME shell only give back the first string element when you request the window class? Sorry, I'm not very familiar with how the extensions work, or what window info the GNOME shell is able to provide to these extensions. 

And of course the closely related question would be, do you also only match application class to this first element of the WM_CLASS pair on X11/Xorg? Did you start out doing that in the beginning, and if so, why? Why not the second element? 

I'm really wondering if xkeysnail (and thus keyszer) made a mistake at some point by deciding to pull the second element for application class matching, rather than using the first element. 

The two strings in WM_CLASS are often identical, or with only case differences, but sometimes they are quite different. So if you've built up a config that matches on what comes back from the second string, it will often fail to match the first string, even with a case insensitive regex. 

There seems to be no such issue with WM_NAME (well, _NET_WM_NAME to avoid legacy encoding issues) in X11/Xorg and the "title" in Wayland being provided by these GNOME extensions, since it's just one string (and the same string) returned in either environment.

**Comment by k0kubun** [comment]
Honestly, I don't remember that much about this problem, which implies that it was probably not a deliberate choice despite your expectation :slightly_smiling_face: 

Full disclosure, this extension itself is a fork of [what powers another tool](https://github.com/iberianpig/fusuma-plugin-appmatcher/blob/af3a709ae3d5278ae62c965d6871c9608135d18a/lib/fusuma/plugin/appmatcher/gnome_extensions/appmatcher%40iberianpig.dev/extension.js) (the author uses xremap and helped me do this). I don't know why you have to maintain an xkeysnail fork when you could just use xremap, but please feel free to fork xremap/xremap-gnome to change that behavior.

**Comment by RedBearAK** [comment]
@k0kubun 

Thanks for the response. And the link. 

> feel free to fork xremap/xremap-gnome to change that behavior

Appreciate that. Fortunately, as long as either of the GNOME extensions in question are properly maintained, we don't need to fork anything. All we have to do is set up a module that knows how to communicate with the extension over D-Bus. And I already have that working, for both extensions. 

> I don't know why you have to maintain an xkeysnail fork when you could just use xremap

It's a matter of having an existing ecosystem. I use a project called Kinto, which is a fairly complex config for xkeysnail to make a large selection of different Linux applications, and general system shortcuts, behave is if you are using a Mac. Basically it makes it so you can go back and forth between a Linux machine and a Mac and see virtually no difference in how keyboard shortcuts work. Not perfect, but it works astonishingly well. Well enough that I've often forgotten whether I was working on my Linux laptop or my MacBook Pro running macOS. 

Converting the Kinto config over to a completely different keymapper might be possible, but it would be a pretty big task. On the other hand, keyszer is basically backwards compatible while adding a lot of new capabilities, like matching on window name. So converting an existing Kinto config over to using keyszer is pretty easy. It's main failing is just that it hasn't been usable outside of X11 until now. 

At the link below is an older version of my own customized Kinto config, with a lot of features beyond what a base Kinto install provides. At the time I submitted this PR, my config was around 3,000 lines, as you can see. 

https://github.com/rbreaves/kinto/pull/750/files

My current Kinto config has evolved quite a lot more than the files at this link. I can't even imagine what it will be like to try and replicate the functionality of this setup with a different keymapper. But it's a possibility for the future. I've definitely been checking out your project, and I think the main dev of Kinto has too. 

Thanks again. 👍🏽

**Comment by k0kubun** [comment]
> Converting the Kinto config over to a completely different keymapper might be possible, but it would be a pretty big task.

Out of curiosity, if somebody steps up to rewrite the xkeysnail config Python script into an xremap config YAML file that works in the same way, would you think it'd be a better solution? (e.g. no interpreter dependency, faster)

We could also make [xremap's Python interface](https://github.com/xremap/xremap-python/blob/master/example/config.py) closer to xkeysnail if they prefer to configure things in Python.

> On the other hand, keyszer is basically backwards compatible while adding a lot of new capabilities, like matching on window name.

I admit that xremap still doesn't support matching window names. Is there anything else you think is missing in xremap compared to your fork? I've supported people who switch from xkeysnail to xremap and added many xkeysnail-like features while doing so. I'd be curious to know what I haven't done yet.

**Comment by RedBearAK** [comment]
> Out of curiosity, if somebody steps up to rewrite the xkeysnail config Python script into an xremap config YAML file that works in the same way, would you think it'd be a better solution? (e.g. no interpreter dependency, faster)

I would be interested in finding out exactly how closely an independent YAML config could replicate what I've got now, which has become quite dynamic in some aspects. 

With the config file for Kinto (while it still uses xkeysnail/keyszer) being straight Python code, I've been able to take advantage of that to do some things that might be difficult with a different kind of config. 

But, even keyszer still has some limitations, that required some "hacks" to get what I wanted out of it. For instance, the `optspecialchars.py` file at that link is a complete implementation of the special character entry scheme available to every user of macOS. It gives access to all the characters documented here: 

https://github.com/RedBearAK/optspecialchars

These are the characters available (583) on the far more extensive "ABC Extended" layout: 

https://github.com/RedBearAK/optspecialchars/blob/main/Option-key-characters_ABC-Extended.md

Characters like this: 

¡™£¢∞§¶•ªº–≠
Œ„´‰ˇÁ¨ˆØ∏”’»
èáöîñÜÅ

The first thing that had to be added to keyszer to support this was a function that would process Unicode addresses, and convert them into "Shift-Ctrl-u, [Unicode address], Enter" sets of keystrokes. This ability was already built in to AHK on Windows. On Linux this has some unfortunate problems, like needing IBus to work (although it might also work with `fcitx`). 

But I also wanted to mimic the special behavior of macOS when you use "dead keys" to create accented characters. This is a key sequence that has different outcomes and behaviors depending what key you press after the initial shortcut. AutoHotkey is very powerful, so doing it on Windows wasn't too difficult. But with keyszer, the dev had to help me figure out how to set up a "tripwire" keymap with a conditional function that gets evaluated at a certain point in the config (on every key press) and makes a necessary thing happen to make it all work. The whole thing was so complicated that the details are actually getting a little hazy. 

> I admit that xremap still doesn't support matching window names. 

That's something that really needs to happen. Matching on WM_NAME has become an essential part of my custom Kinto config. I've used that feature after keyszer implemented it a few months ago, to solve some sticky problems. Mostly with how Cmd+W behaves with a lot of Linux application windows. Or making some shortcuts work correctly in some odd child windows: 

```py
# Keybindings overrides for GNOME XDG "Save As" and "Open File" dialogs
file_open_save_dialogs = [
    {cls:"^xdg-desktop-portal-gnome$|^firefox.*$", nme:"^Open File$"},
    {cls:"^xdg-desktop-portal-gnome$|^firefox.*$", nme:"^Save As$"},
]
keymap("XDG file dialogs", {
    C("RC-Left"):               C("Alt-Left"),                  # Go Back
    C("RC-Right"):              C("Alt-Right"),                 # Go Forward
    C("RC-Up"):                 C("Alt-Up"),                    # Go Up dir
    C("RC-Down"):               C("Enter"),                     # Go Down dir (open folder/file) [universal]
}, when = matchProps(lst=file_open_save_dialogs))
```

I've seen other people talking about maybe doing special shortcuts for specific websites by matching on the Firefox or Chrome tab names, which is in the WM_NAME attribute. But I haven't attempted anything like that. 

Actually I've been a little surprised at how limited the number of situations are where I've found a use for WM_NAME. But when you do need it, you really need it. 

> Is there anything else you think is missing in xremap compared to your fork? I've supported people who switch from xkeysnail to xremap and added many xkeysnail-like features while doing so. I'd be curious to know what I haven't done yet.

Not sure yet. I'll have to get more familiar with what xremap can or can't do in its current form, to know how to compare it with keyszer. 

I wanted to do a thing with the Enter key in Linux file managers, to make it work like the Finder in macOS. Unlike Windows and Linux, macOS has always used the Enter key for renaming files/folders, instead of opening/launching. You go "Enter+rename+Enter." It's very quick. Once I got used to this, it was very irritating going back to Windows or Linux file managers and having to attempt to use F2 (often hijacked for some hardware function on PC laptops) or the infernal right-click menu, just to rename a file. 

So eventually I made a function that causes the Enter key to behave like it does in the Finder, whenever I'm in any common Linux file manager like Nautilus or Dolphin. 

```py
# Boolean variable to toggle Enter key state between F2 and Enter
# True = Enter key sends F2, False = Enter key sends Enter
_enter_is_F2 = True # DON'T CHANGE THIS! Must be set to True here. 


def is_Enter_F2(combo_if_true, combo_if_false):
    """
    Send a different combo for Enter key depending on state of _enter_is_F2 variable,\n 
    or latch the variable to True or False to control the Enter key state on next use.
    """
    def _is_Enter_F2():
        global _enter_is_F2
        combo_list = [combo_if_true]
        if combo_if_false in (True, False):                     # Latch variable to given bool value
            _enter_is_F2 = combo_if_false
        elif _enter_is_F2:                                      # If Enter is F2 now, set to be Enter next
            _enter_is_F2 = False
        else:                                                   # If Enter is Enter now, set to be F2 next
            combo_list = [combo_if_false]
            _enter_is_F2 = True
        return combo_list
    return _is_Enter_F2
```

And then it's controlled with mappings like this: 

```py
    C("Enter"):                 is_Enter_F2(C("F2"),C("Enter")),        # Send F2 to rename files, unless var is False
    C("Shift-RC-N"):            is_Enter_F2(C("Shift-RC-N"), False),    # New folder, set Enter to Enter
    C("RC-L"):                  is_Enter_F2(C("RC-L"), False),          # Set Enter to Enter for Location field
    C("RC-F"):                  is_Enter_F2(C("RC-F"), False),          # Set Enter to Enter for Find field
    C("Esc"):                   is_Enter_F2(C("Esc"), True),            # Send Escape, make sure Enter is back to F2
```

What you may notice here is that other keys need to be able to control the state of the Enter key being Enter or F2. In Windows with AHK there is knowledge of the exact part of the Explorer window you're in, so you can have the Enter key go back to being Enter in text areas. In Linux there is no such information available, so I had to fake it. For instance, when you use the Enter key, it sends out F2 but toggles its own state, so when you're finished renaming the file and hit Enter, it's actually the Enter key. It's not perfect but works pretty well. 

Now, if there were a way to do the kinds of things described above more easily with xremap, you'd certainly have my attention. 

There's also a string processor in keyszer now, which uses a Unicode processor function to deal with anything outside of basic ASCII characters. I can take a small amount of credit for getting those functions developed. Right now it has a problem with doing the wrong thing when CapsLock is on, but I've already put together a fix for that, that I hope will be merged soon. 

It lets you take a crazy string like this: 

```py
ST("CapsLock 🌹 € ÿ \u2021 \U00002021 12345 !@#$% |\ double-tapped.")
```

And get this: 

CapsLock 🌹 € ÿ ‡ ‡ 12345 !@#$% |\ double-tapped.

All "typed" out with keystrokes from the virtual keyboard. 

Oh, and I wanted to do special things with my numpad keys when NumLock was either on or off, so I figured out how to add NumLock and CapsLock as available context properties. So keyszer actually has five different properties available for conditions to attach to keymaps or modmaps: 

- Window class
- Window name/title
- NumLock LED state
- CapsLock LED state
- Device name

Why I needed NumLock state is a long story involving buggy GTK3 applications and a laptop keyboard that has no "real" PgUp/PgDn/Home/End keys anywhere, only the keypad versions, which are different key codes. 

And then I wrote a rather ridiculous function that allows matching on all these properties (simultaneously if you want) in various different ways to suit nearly any purpose. 

https://github.com/joshgoebel/keyszer/wiki/Helper-Function-Examples#function-matchprops

Let's see, and just in the last week or so I've figured out how to automate the modmaps so that the Kinto config will just automatically work with any of the different keyboard types it supports (IBM, Chromebook, Windows, Apple) without the user needing to manually switch the keyboard type. This will even allow the use of more than one keyboard type connected to the same machine at the same time. Which is pretty neat, since a lot of what makes Kinto special is the way it moves the modifier keys around for each keyboard type differently, so they all end up working like a Mac. 

Anyway, this is probably much more than you wanted to hear about. The main thing I appreciate is that you're interested in adding missing features to xremap, if there are any. I'm sure we can talk more about this at some point. 👍🏽

**Comment by k0kubun** [comment]
Thanks for the details :eyes: I'll revisit this discussion and address the differences later. Reopening this as a reminder for me.

**Comment by plundration** [comment]
Would love to be able to use accents (dead keys and accented letters). Coming from xmodmap.
